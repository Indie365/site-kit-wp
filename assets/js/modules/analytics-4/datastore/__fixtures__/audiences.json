[
    {
        "adsPersonalizationEnabled": true,
        "description": "All users",
        "displayName": "All Users",
        "exclusionDurationMode": null,
        "membershipDurationDays": 540,
        "name": "properties/12345/audiences/1"
    },
    {
        "adsPersonalizationEnabled": true,
        "description": "Users who have made a purchase",
        "displayName": "Purchasers",
        "exclusionDurationMode": null,
        "membershipDurationDays": 540,
        "name": "properties/12345/audiences/2"
    },
    {
        "adsPersonalizationEnabled": true,
        "description": "Description",
        "displayName": "Test Audience",
        "exclusionDurationMode": null,
        "membershipDurationDays": 30,
        "name": "properties/12345/audiences/3",
        "filterClauses": [
            {
				"clauseType": "INCLUDE",
				"simpleFilter": {
					"scope": "AUDIENCE_FILTER_SCOPE_ACROSS_ALL_SESSIONS",
					"filterExpression": {
                        "andGroup": {
                            "filterExpressions": [
                                {
                                    "orGroup": {
                                        "filterExpressions": [
                                            {
                                                "dimensionOrMetricFilter": {
                                                    "atAnyPointInTime": null,
                                                    "fieldName": "newVsReturning",
                                                    "inAnyNDayPeriod": null,
                                                    "stringFilter": {
                                                        "caseSensitive": null,
                                                        "matchType": "EXACT",
                                                        "value": "new"
                                                    }
                                                }
                                            }
                                        ]
                                    }
                                }
                            ]
                        }
                    }
				}
			}
        ]
    }
]