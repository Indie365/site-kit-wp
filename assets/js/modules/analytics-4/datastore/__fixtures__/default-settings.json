{
  "ownerID": 0,
  "propertyID": "",
  "webDataStreamID": "",
  "measurementID": "",
  "useSnippet": true
}
