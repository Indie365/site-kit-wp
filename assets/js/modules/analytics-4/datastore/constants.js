/**
 * `modules/analytics-4` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const MODULES_ANALYTICS_4 = 'modules/analytics-4';

export const PROPERTY_CREATE = 'property_create';
export const WEBDATASTREAM_CREATE = 'webdatastream_create';

export const MAX_WEBDATASTREAMS_PER_BATCH = 10;

// Date range offset days for Analytics 4 report requests.
export const DATE_RANGE_OFFSET = 1;

export const GTM_SCOPE = 'https://www.googleapis.com/auth/tagmanager.readonly';

export const ENHANCED_MEASUREMENT_FORM = 'enhanced-measurement-form';
export const ENHANCED_MEASUREMENT_ENABLED = 'enhanced-measurement-enabled';
export const ENHANCED_MEASUREMENT_SHOULD_DISMISS_ACTIVATION_BANNER =
	'enhanced-measurement-should-dismiss-activation-banner';

export const FORM_CUSTOM_DIMENSIONS_CREATE = 'analytics4CustomDimensionsCreate';

export const CUSTOM_DIMENSION_DEFINITIONS = {
	googlesitekit_post_date: {
		parameterName: 'googlesitekit_post_date',
		displayName: __( 'WordPress Post Creation Date', 'google-site-kit' ),
		description: __(
			'Date of which this post was published',
			'google-site-kit'
		),
		scope: 'EVENT',
	},
	googlesitekit_post_author: {
		parameterName: 'googlesitekit_post_author',
		displayName: __( 'WordPress Post Author', 'google-site-kit' ),
		description: __(
			'User ID of the author for this post',
			'google-site-kit'
		),
		scope: 'EVENT',
	},
	googlesitekit_post_categories: {
		parameterName: 'googlesitekit_post_categories',
		displayName: __( 'WordPress Post Categories', 'google-site-kit' ),
		description: __(
			'Comma-separated list of category IDs assigned to this post',
			'google-site-kit'
		),
		scope: 'EVENT',
	},
	googlesitekit_post_type: {
		parameterName: 'googlesitekit_post_type',
		displayName: __( 'WordPress Post Type', 'google-site-kit' ),
		description: __( 'Content type for this post', 'google-site-kit' ),
		scope: 'EVENT',
	},
};
