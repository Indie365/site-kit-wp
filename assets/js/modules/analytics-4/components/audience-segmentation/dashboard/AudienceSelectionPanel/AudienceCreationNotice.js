/**
 * Audience Selection Panel AudienceCreationNotice component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { AUDIENCE_CREATION_NOTICE_SLUG } from './constants';
import { useDispatch, useSelect } from 'googlesitekit-data';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import {
	MODULES_ANALYTICS_4,
	SITE_KIT_AUDIENCE_DEFINITIONS,
} from '../../../../datastore/constants';
import Link from '../../../../../../components/Link';
import CloseIcon from '../../../../../../../svg/icons/close.svg';
import SpinnerButton, {
	SPINNER_POSITION,
} from '../../../../../../googlesitekit/components-gm2/SpinnerButton';

export default function AudienceCreationNotice() {
	const availableAudiences = useSelect( ( select ) => {
		const { getConfigurableAudiences } = select( MODULES_ANALYTICS_4 );

		const audiences = getConfigurableAudiences();

		if ( undefined === audiences ) {
			return undefined;
		}

		if ( ! audiences.length ) {
			return [];
		}

		return audiences.filter(
			( { audienceType } ) => audienceType === 'SITE_KIT_AUDIENCE'
		);
	} );

	const { dismissItem } = useDispatch( CORE_USER );

	const isDismissed = useSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed( AUDIENCE_CREATION_NOTICE_SLUG )
	);

	const onCloseClick = () => {
		dismissItem( AUDIENCE_CREATION_NOTICE_SLUG );
	};
	if (
		! availableAudiences ||
		availableAudiences.length === 0 ||
		isDismissed
	) {
		return null;
	}

	return (
		<div className="googlesitekit-audience-selection-panel__audience-creation-notice">
			<div className="googlesitekit-audience-selection-panel__audience-creation-notice-header">
				<p className="googlesitekit-audience-selection-panel__audience-creation-notice-title">
					{ __(
						'Create groups suggested by Site Kit.',
						'google-site-kit'
					) }
				</p>

				<Link
					className="googlesitekit-audience-selection-panel__audience-creation-notice-close"
					onClick={ onCloseClick }
					linkButton
				>
					<CloseIcon width="15" height="15" />
				</Link>
			</div>
			<div className="googlesitekit-audience-selection-panel__audience-creation-notice-body">
				{ availableAudiences &&
					availableAudiences.map( ( audience ) => (
						<div
							key={ audience.name }
							className="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
						>
							<div className="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details">
								<h3>{ audience.displayName }</h3>
								<p className="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description">
									{
										SITE_KIT_AUDIENCE_DEFINITIONS[
											audience.audienceSlug
										].description
									}
								</p>
							</div>
							<div className="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button">
								<SpinnerButton
									spinnerPosition={ SPINNER_POSITION.BEFORE }
									onClick={ () => {
										// TODO: to be implemented in #8164
									} }
								>
									{ __( 'Create', 'google-site-kit' ) }
								</SpinnerButton>
							</div>
						</div>
					) ) }
			</div>
		</div>
	);
}
