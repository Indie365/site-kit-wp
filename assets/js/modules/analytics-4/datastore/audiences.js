/**
 * `modules/analytics-4` data store: audiences.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import Data from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from './constants';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { createValidatedAction } from '../../../googlesitekit/data/utils';
import { validateAudience } from '../utils/validation';

const fetchGetAudiencesStore = createFetchStore( {
	baseName: 'getAudiences',
	controlCallback() {
		return API.get(
			'modules',
			'analytics-4',
			'audiences',
			{},
			{
				useCache: false,
			}
		);
	},
	reducerCallback( state, audiencesResponse ) {
		return { ...state, audiences: audiencesResponse.audiences };
	},
} );

const fetchCreateAudienceStore = createFetchStore( {
	baseName: 'createAudience',
	controlCallback: ( { audience } ) =>
		API.set( 'modules', 'analytics-4', 'create-audience', {
			audience,
		} ),
	reducerCallback( state, audience ) {
		return {
			...state,
			audiences: [ ...( state.audiences || [] ), audience ],
		};
	},
	argsToParams: ( audience ) => ( {
		audience,
	} ),
	validateParams: ( { audience } ) => {
		validateAudience( audience );
	},
} );

const baseInitialState = {
	audiences: undefined,
};

const baseActions = {
	/**
	 * Creates new property audience.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} audience                             The property audience parameters.
	 * @param {string} [audience.displayName]               Required. The display name of the Audience.
	 * @param {string} [audience.description]               Required. The description of the Audience.
	 * @param {number} [audience.membershipDurationDays]    Required. The duration a user should stay in an Audience. Cannot be more than 540 days.
	 * @param {Array}  [audience.filterClauses]             Required. Filter clauses array of <AudienceFilterClause> objects that define the Audience.
	 * @param {Object} [audience.eventTrigger]              Optional. Specifies an event to log when a user joins the Audience.
	 * @param {string} [audience.eventTrigger.eventName]    Required if `eventTrigger` is provided. The event name that will be logged.
	 * @param {string} [audience.eventTrigger.logCondition] Required if `eventTrigger` is provided. When to log the event. Acceptable values:
	 *                                                      - 'LOG_CONDITION_UNSPECIFIED': Log condition is not specified.
	 *                                                      - 'AUDIENCE_JOINED': The event should be logged only when a user is joined.
	 *                                                      - 'AUDIENCE_MEMBERSHIP_RENEWED': The event should be logged whenever the Audience condition is met, even if the user is already a member of the Audience.
	 * @param {string} [audience.exclusionDurationMode]     Optional. Specifies how long an exclusion lasts for users that meet the exclusion filter. Acceptable values:
	 *                                                      - 'AUDIENCE_EXCLUSION_DURATION_MODE_UNSPECIFIED': Not specified.
	 *                                                      - 'EXCLUDE_TEMPORARILY': Exclude users from the Audience during periods when they meet the filter clause.
	 *                                                      - 'EXCLUDE_PERMANENTLY': Exclude users from the Audience if they've ever met the filter clause.
	 * @return {Object} Object with `response` and `error`.
	 */
	createAudience: createValidatedAction(
		validateAudience,
		function* ( audience ) {
			const { response, error } =
				yield fetchCreateAudienceStore.actions.fetchCreateAudience(
					audience
				);

			return { response, error };
		}
	),
};

const baseControls = {};

const baseReducer = ( state, { type } ) => {
	switch ( type ) {
		default: {
			return state;
		}
	}
};

const baseResolvers = {
	*getAudiences() {
		const registry = yield Data.commonActions.getRegistry();

		const audiences = registry.select( MODULES_ANALYTICS_4 ).getAudiences();

		if ( audiences === undefined ) {
			yield fetchGetAudiencesStore.actions.fetchGetAudiences();
		}
	},
};

const baseSelectors = {
	/**
	 * Gets the property audiences.
	 *
	 * @since n.e.x.t
	 *
	 * @param {Object} state Data store's state.
	 * @return {(Array|undefined)} An array with audiences objects; `undefined` if not loaded.
	 */
	getAudiences( state ) {
		return state.audiences;
	},
};

const store = Data.combineStores(
	fetchGetAudiencesStore,
	fetchCreateAudienceStore,
	{
		initialState: baseInitialState,
		actions: baseActions,
		controls: baseControls,
		reducer: baseReducer,
		resolvers: baseResolvers,
		selectors: baseSelectors,
	}
);

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
