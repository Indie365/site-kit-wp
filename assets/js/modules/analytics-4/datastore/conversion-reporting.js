/**
 * `modules/analytics-4` data store: conversion reporting.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { createRegistrySelector } from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from './constants';

export const selectors = {
	/**
	 * Checks whether the provided conversion reporting events are available.
	 *
	 * @since 1.135.0
	 *
	 * @param {Object}               state  Data store's state.
	 * @param {string|Array<string>} events Conversion reporting events to check.
	 * @return {(boolean|undefined)} True if all provided custom dimensions are available, otherwise false. Undefined if available custom dimensions are not loaded yet.
	 */
	hasConversionReportingEvents: createRegistrySelector(
		( select ) => ( state, events ) => {
			// Ensure events is always an array, even if a string is passed.
			const eventsToCheck = Array.isArray( events ) ? events : [ events ];

			const detectedEvents =
				select( MODULES_ANALYTICS_4 ).getDetectedEvents();

			if ( ! detectedEvents?.length ) {
				return false;
			}

			return eventsToCheck.some( ( event ) =>
				detectedEvents.includes( event )
			);
		}
	),
};

export default {
	selectors,
};
