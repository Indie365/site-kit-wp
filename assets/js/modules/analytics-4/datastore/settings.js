/**
 * `modules/analytics-4` data store: settings.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import { createStrictSelect } from '../../../googlesitekit/data/utils';
import {
	isValidPropertySelection,
	isValidWebDataStreamID,
	isValidWebDataStreamSelection,
} from '../utils/validation';
import {
	INVARIANT_DOING_SUBMIT_CHANGES,
	INVARIANT_SETTINGS_NOT_CHANGED,
} from '../../../googlesitekit/data/create-settings-store';
import { MODULES_ANALYTICS } from '../../analytics/datastore/constants';
import {
	MODULES_ANALYTICS_4,
	PROPERTY_CREATE,
	WEBDATASTREAM_CREATE,
} from './constants';

// Invariant error messages.
export const INVARIANT_INVALID_PROPERTY_SELECTION =
	'a valid propertyID is required to submit changes';
export const INVARIANT_INVALID_WEBDATASTREAM_ID =
	'a valid webDataStreamID is required to submit changes';

export async function submitChanges( { select, dispatch } ) {
	let propertyID = select( MODULES_ANALYTICS_4 ).getPropertyID();
	if ( propertyID === PROPERTY_CREATE ) {
		const accountID = select( MODULES_ANALYTICS ).getAccountID();
		const { response: property, error } = await dispatch(
			MODULES_ANALYTICS_4
		).createProperty( accountID );
		if ( error ) {
			return { error };
		}

		propertyID = property._id;
		dispatch( MODULES_ANALYTICS_4 ).setPropertyID( propertyID );
		dispatch( MODULES_ANALYTICS_4 ).setWebDataStreamID(
			WEBDATASTREAM_CREATE
		);
		dispatch( MODULES_ANALYTICS_4 ).setMeasurementID( '' );
	}

	const webDataStreamID = select( MODULES_ANALYTICS_4 ).getWebDataStreamID();
	if (
		propertyID &&
		( webDataStreamID === WEBDATASTREAM_CREATE ||
			! isValidWebDataStreamID( webDataStreamID ) )
	) {
		const { response: webdatastream, error } = await dispatch(
			MODULES_ANALYTICS_4
		).createWebDataStream( propertyID );
		if ( error ) {
			return { error };
		}

		dispatch( MODULES_ANALYTICS_4 ).setWebDataStreamID( webdatastream._id );
		dispatch( MODULES_ANALYTICS_4 ).setMeasurementID(
			// eslint-disable-next-line sitekit/acronym-case
			webdatastream.measurementId
		);
	}

	if ( select( MODULES_ANALYTICS_4 ).haveSettingsChanged() ) {
		const { error } = await dispatch( MODULES_ANALYTICS_4 ).saveSettings();
		if ( error ) {
			return { error };
		}
	}

	await API.invalidateCache( 'modules', 'analytics-4' );

	return {};
}

export function validateCanSubmitChanges( select ) {
	if ( select( MODULES_ANALYTICS_4 ).isAdminAPIWorking() === false ) {
		return;
	}

	const {
		haveSettingsChanged: haveGA4SettingsChanged,
		isDoingSubmitChanges,
		getPropertyID,
		getWebDataStreamID,
	} = createStrictSelect( select )( MODULES_ANALYTICS_4 );

	const { haveSettingsChanged: haveUASettingsChanged } = createStrictSelect(
		select
	)( MODULES_ANALYTICS );

	// Check if we have GA4 settings changed only if we are sure that there is no UA changes.
	if ( ! haveUASettingsChanged() ) {
		invariant( haveGA4SettingsChanged(), INVARIANT_SETTINGS_NOT_CHANGED );
	}

	invariant( ! isDoingSubmitChanges(), INVARIANT_DOING_SUBMIT_CHANGES );

	const propertyID = getPropertyID();
	invariant(
		isValidPropertySelection( propertyID ),
		INVARIANT_INVALID_PROPERTY_SELECTION
	);
	if ( propertyID !== PROPERTY_CREATE ) {
		invariant(
			isValidWebDataStreamSelection( getWebDataStreamID() ),
			INVARIANT_INVALID_WEBDATASTREAM_ID
		);
	}
}
