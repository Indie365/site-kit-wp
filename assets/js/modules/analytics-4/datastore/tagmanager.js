/**
 * `modules/analytics-4` data store: tagmanager.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { CORE_MODULES } from '../../../googlesitekit/modules/datastore/constants';
import { MODULES_TAGMANAGER } from '../../tagmanager/datastore/constants';

const { createRegistrySelector } = Data;

const baseSelectors = {
	/**
	 * Determines whether the live container version has finished loading.
	 *
	 * @since n.e.x.t
	 *
	 * @return {boolean} TRUE if the GTM module is not available or the live container version has finished loading, otherwise FALSE.
	 */
	hasFinishedLoadingGTMContainers: createRegistrySelector(
		( select ) => () => {
			const tagmanagerModuleConnected =
				select( CORE_MODULES ).isModuleConnected( 'tagmanager' );
			if ( ! tagmanagerModuleConnected ) {
				return true;
			}

			const accountID = select( MODULES_TAGMANAGER ).getAccountID();
			const internalContainerID =
				select( MODULES_TAGMANAGER ).getInternalContainerID();
			const internalAMPContainerID =
				select( MODULES_TAGMANAGER ).getInternalAMPContainerID();

			return (
				select( MODULES_TAGMANAGER ).hasFinishedResolution(
					'getLiveContainerVersion',
					[ accountID, internalContainerID ]
				) ||
				select( MODULES_TAGMANAGER ).hasFinishedResolution(
					'getLiveContainerVersion',
					[ accountID, internalAMPContainerID ]
				)
			);
		}
	),
};

const store = {
	selectors: baseSelectors,
};

export default store;
