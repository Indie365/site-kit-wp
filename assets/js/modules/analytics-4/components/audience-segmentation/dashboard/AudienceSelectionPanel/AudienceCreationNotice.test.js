/**
 * Audience Selection Panel AudienceCreationNotice tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	provideModules,
	provideModuleRegistrations,
	provideUserInfo,
} from '../../../../../../../../tests/js/test-utils';
import { CORE_USER } from '../../../../../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../../../datastore/constants';
import AudienceCreationNotice from './AudienceCreationNotice';
import { AUDIENCE_CREATION_NOTICE_SLUG } from './constants';
import { availableAudiences } from '../../../../datastore/__fixtures__';

describe( 'AudienceCreationNotice', () => {
	let registry;

	const fetchGetDismissedItemsRegExp = new RegExp(
		'^/google-site-kit/v1/core/user/data/dismissed-items'
	);

	beforeEach( () => {
		registry = createTestRegistry();
		provideUserInfo( registry );
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics-4',
			},
		] );
		provideModuleRegistrations( registry );

		// provideUserInfo( registry );
		// provideModules( registry );
		// provideModuleRegistrations( registry );
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );

		fetchMock.getOnce( fetchGetDismissedItemsRegExp, { body: [] } );
	} );

	it( 'should render null if no audiences are available', () => {
		const { container } = render( <AudienceCreationNotice />, {
			registry,
		} );

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render null if there are 0 audiences available', async () => {
		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			availableAudiences: [],
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render null if the user has dismissed the notice', async () => {
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [ AUDIENCE_CREATION_NOTICE_SLUG ] );

		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences,
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toBeEmptyDOMElement();
	} );

	it( 'should render the notice if the user has not dismissed the notice', async () => {
		registry
			.dispatch( MODULES_ANALYTICS_4 )
			.receiveResourceDataAvailabilityDates( {
				audience: availableAudiences.reduce( ( acc, { name } ) => {
					acc[ name ] = 20201220;

					return acc;
				}, {} ),
				customDimension: {},
				property: {},
			} );

		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
			accountID: '12345',
			propertyID: '34567',
			measurementID: '56789',
			webDataStreamID: '78901',
			availableAudiences,
		} );

		const { container, waitForRegistry } = render(
			<AudienceCreationNotice />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );
} );
