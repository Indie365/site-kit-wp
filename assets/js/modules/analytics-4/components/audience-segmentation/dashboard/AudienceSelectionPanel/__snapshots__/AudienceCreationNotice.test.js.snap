// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceCreationNotice should render the notice if the user has not dismissed the notice 1`] = `
<div>
  <div
    class="googlesitekit-audience-selection-panel__audience-creation-notice"
  >
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-header"
    >
      <p
        class="googlesitekit-audience-selection-panel__audience-creation-notice-title"
      >
        Create groups suggested by Site Kit.
      </p>
      <button
        class="googlesitekit-cta-link googlesitekit-audience-selection-panel__audience-creation-notice-close googlesitekit-cta-link--link-button"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          <svg />
        </span>
      </button>
    </div>
    <div
      class="googlesitekit-audience-selection-panel__audience-creation-notice-body"
    >
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            New visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who visited the site for the first time
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
      <div
        class="googlesitekit-audience-selection-panel__audience-creation-notice-audience"
      >
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-details"
        >
          <h3>
            Returning visitors
          </h3>
          <p
            class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-description"
          >
            People who have visited your site at least once before
          </p>
        </div>
        <div
          class="googlesitekit-audience-selection-panel__audience-creation-notice-audience-button"
        >
          <button
            class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__before mdc-button--raised"
            target="_self"
          >
            <span
              class="mdc-button__label"
            >
              Create
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
